tensorflow_probability/python/experimental/bijectors/highway_flow_test.py

```diff
@@ -0,0 +1,149 @@
+
+    bijector = tfp.experimental.bijectors.build_highway_flow_layer(
+      width, activation_fn=False)
+    self.evaluate(
+      [v.initializer for v in bijector.trainable_variables])
+    self.assertStartsWith(bijector.name, 'highway_flow')
+    self.assertAllClose(x, bijector.inverse(
+      tf.identity(bijector.forward(x))))
+    self.assertAllClose(
+      bijector.forward_log_det_jacobian(x, event_ndims=2),
+      -bijector.inverse_log_det_jacobian(
+        tf.identity(bijector.forward(x)), event_ndims=2))
+
+  def testGating(self):
+    width = 4
+    x = tf.ones((2, width,
+                 width)) * samplers.uniform((2, width, width),
+                                            minval=-1.,
+                                            maxval=1., seed=test_util.test_seed(
+        sampler_type='stateless'))
+
+    # Test with gating half of the inputs
+    bijector = tfp.experimental.bijectors.build_highway_flow_layer(
+      width, activation_fn=True, gate_first_n=2)
+    self.evaluate(
+      [v.initializer for v in bijector.trainable_variables])
+    self.assertStartsWith(bijector.name, 'highway_flow')
+    self.assertAllClose(x, bijector.inverse(
+      tf.identity(bijector.forward(x))))
+    self.assertAllClose(
+      bijector.forward_log_det_jacobian(x, event_ndims=2),
+      -bijector.inverse_log_det_jacobian(
+        tf.identity(bijector.forward(x)), event_ndims=2))
+
+    # Test with gating no inputs
+    bijector = tfp.experimental.bijectors.build_highway_flow_layer(
+      width, activation_fn=True, gate_first_n=0)
+    self.evaluate(
+      [v.initializer for v in bijector.trainable_variables])
+    self.assertStartsWith(bijector.name, 'highway_flow')
+    self.assertAllClose(x, bijector.inverse(
+      tf.identity(bijector.forward(x))))
+    self.assertAllClose(
+      bijector.forward_log_det_jacobian(x, event_ndims=2),
+      -bijector.inverse_log_det_jacobian(
+        tf.identity(bijector.forward(x)), event_ndims=2))
+
+  def testResidualFractionGradientsWithCenteredDifference(self):
+    width = 4
+    batch_size = 3
+    residual_fraction = tf.constant(0.5)
+    bijector = tfp.experimental.bijectors.HighwayFlow(
+      residual_fraction=residual_fraction,
+      activation_fn=tf.nn.softplus,
+      bias=tf.zeros(width),
+      upper_diagonal_weights_matrix=tf.eye(width),
+      lower_diagonal_weights_matrix=tf.eye(width),
+      gate_first_n=width
+    )
+    target = tfd.MultivariateNormalDiag(loc=tf.zeros(width),
+                                        scale_diag=tf.ones(width))
+    x = tf.ones((batch_size, width))
+    with tf.GradientTape() as g:
+      g.watch(bijector.residual_fraction)
+      y = tf.reduce_mean(target.log_prob(bijector.forward(x)))
+    tf_grad = g.gradient(y, bijector.residual_fraction)
+
+    h = 1e-3
+
+    # pylint: disable=protected-access
+    bijector._residual_fraction = residual_fraction + h
+    y1 = tf.reduce_mean(target.log_prob(bijector.forward(x)))
```

Comment:
I think the bijector cache will give a wrong answer here (and two lines below), since it doesn't know that `_residual_fraction` has changed from the previous call. You'll probably need to pass `tf.identity(x)`.